plugins/woocommerce-blocks/assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,140 +1,319 @@
+	 * Specifically manages transitions to and from single product templates.
+	 *
+	 * @param {string | undefined} previousTemplateId - The previous template ID
+	 */
+	private handleTemplateChange(
+		previousTemplateId: string | undefined
+	): void {
+		const isTransitioningToOrFromSingleProduct =
+			this.currentTemplateId?.includes( 'single-product' ) ||
+			previousTemplateId?.includes( 'single-product' );
+
+		if ( ! isTransitioningToOrFromSingleProduct ) {
+			return;
+		}
+
+		this.blocks.forEach( ( config ) => {
+			if ( ! this.isRegistering ) {
+				this.unregisterBlock( config );
+				this.registerBlock( config );
 			}
-		}, 'core/edit-site' );
+		} );
 	}
 
-	if ( isAvailableOnPostEditor ) {
-		subscribe( () => {
-			const isRegistered = Boolean( variationName )
-				? blockRegistrationAttempts.has( variationName )
-				: blockRegistrationAttempts.has( blockName ) ||
-				  isBlockRegistered( blockName );
-			// This subscribe callback could be invoked with the core/blocks store
-			// which would cause infinite registration loops because of the `registerBlockType` call.
-			// This local cache helps prevent that.
-			if (
-				! isRegistered &&
-				isAvailableOnPostEditor &&
-				! editSiteStore
-			) {
-				if ( isVariationBlock ) {
-					blockRegistrationAttempts.add( variationName );
-					registerBlockVariation(
-						blockName,
-						blockSettings as BlockVariation< BlockAttributes >
-					);
-				} else {
-					blockRegistrationAttempts.add( blockName );
-					// @ts-expect-error: `registerBlockType` is typed in WordPress core
-					registerBlockType( blockMetadata, blockSettings );
-				}
+	/**
+	 * Unregisters a block or block variation.
+	 * Handles both regular blocks and variations with error handling.
+	 *
+	 * @param {BlockConfig} config - Configuration of the block to unregister
+	 */
+	private unregisterBlock( config: BlockConfig ): void {
+		const { blockName, isVariationBlock, variationName } = config;
+
+		try {
+			if ( isVariationBlock && variationName ) {
+				unregisterBlockVariation( blockName, variationName );
+				this.registeredBlocks.delete( variationName );
+			} else {
+				unregisterBlockType( blockName );
+				this.registeredBlocks.delete( blockName );
 			}
-		}, 'core/edit-post' );
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.debug(
+				`Failed to unregister block ${ blockName }:`,
+				error
+			);
+		}
+	}
+
+	/**
+	 * Registers a block or block variation.
+	 * Handles different registration requirements for various contexts.
+	 * Includes checks to prevent recursive registration.
+	 *
+	 * @param {BlockConfig} config - Configuration of the block to register
+	 */
+	private registerBlock( config: BlockConfig ): void {
+		const {
+			blockName,
+			blockMetadata = blockName,
+			blockSettings,
+			isVariationBlock,
+			variationName,
+			isAvailableOnPostEditor,
+		} = config;
+
+		if ( this.isRegistering ) {
+			return;
+		}
```

Comment:
Is there a particular reason why this variable is needed?

Reply:
It is a protective measure to stop duplicate and recursive registrations

For example, without this flag:
1. A template change triggers block registration
2. During registration, another store update occurs
3. That update triggers another registration attempt
4. We end up in a loop of registration attempts

Reply:
I think this line is enough to cover a loop of registration attempts

```
if ( this.registeredBlocks.has( key ) ) {
    return;
}
```

Reply:
Hmm, I see what you mean but I think `this.isRegistering` serves a different purpose here.

`this.registeredBlocks.has(key)` checks if we've ever attempted to register this specific block before, preventing duplicate attempts for the same block.

`this.isRegistering` protects against recursive registration attempts during the registration process itself. For example:

* We start registering block A
* During registration, a store update triggers another registration attempt
* Without the flag, we could end up with nested registration calls even for different blocks

Reply:
I also want to add I like how explicit the variable name is too.

Reply:
Thinking about this, if block A sets the variable to `true` and block B tries to register it could be prevented from doing so. Let me look into this.

Reply:
Can you review this commit and let me know what you think @gigitux https://github.com/woocommerce/woocommerce/pull/53788/commits/f2851227cd75ee59f625a2a885d87b7228a064f0